### ISSUE TITLE
Unbuffered reading / writing

### ISSUE THREAD

**Issue by benjaoming** [learningequality/ka-lite#2909]
Incidentally, I just merged this in django-dbbackup... it's about writing in chunks: https://bitbucket.org/mjs7231/django-dbbackup/commits/c7e0588c3cc5acd3a96bf7708871edb025b859bc

The problem is what seems to be a common pattern in kalite: Doing `read()` calls without any particular concern of filesize. This operation will load a whole file into memory. But the thing is that we don't know the memory of the target system. In case of Sandisks, Arduinos, Raspberries etc.. it can become problematic.

I suggest having a look at all occurrences of `read\(.*\)`. I haven't seen examples of the same with `write(unbuffered_data)` but it could be a good idea to investigate, too.

**Comment by aronasorman** [comment]
I agree. Since we're gonna do a whole bunch of replacements to solve this issue, the first step would be to make a function that accepts a file-like object and returns a generator for the chunks of each file.

**Comment by benjaoming** [comment]
That's a good idea. That way, depending on the device, we can have a global setting for buffer size.

**Comment by rtibbles** [comment]
This should be less of an issue with content packs now. Punting.

**Comment by shreab373** [comment]
Hi, can I take this? Could you tell me what exactly needs to be done to solve this?

**Comment by benjaoming** [comment]
`retrievecontentpack` only uses calls to `shutil.copyfileobj` and `zipfile.extractall`

Our video download writes in chunks in `fle_utils.internet.download.download_file`

The only problematic thing remaining could be this check on file size from `download_video` in `fle_utils.video`:

```
not len(open(filepath, "rb").read()) == int(response.headers['content-length']))
```

**Comment by benjaoming** [comment]
Fixed in #5436
